hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/observer/ObservableImpl.java

```diff
@@ -0,0 +1,343 @@
+     * is the entity populating the {@link Observable} with data).
+     */
+    public static final String OWNED_OBSERVABLE = "owned_observable";
+
+    private final String name;
+    private final HazelcastInstance hzInstance;
+    private final Consumer<Observable<T>> onDestroy;
+    private final ILogger logger;
+
+    private ConcurrentMap<UUID, RingbufferListener<T>> listeners;
+
+    public ObservableImpl(String name, HazelcastInstance hzInstance, Consumer<Observable<T>> onDestroy, ILogger logger) {
+        this.name = name;
+        this.hzInstance = hzInstance;
+        this.onDestroy = onDestroy;
+        this.logger = logger;
+        this.listeners = new ConcurrentHashMap<>();
+    }
+
+    @Nonnull @Override
+    public String name() {
+        return name;
+    }
+
+    @Nonnull @Override
+    public UUID addObserver(@Nonnull Observer<T> observer) {
+        checkIfLive();
+        UUID id = UuidUtil.newUnsecureUUID();
+        RingbufferListener<T> listener = new RingbufferListener<>(name, id, observer, hzInstance, logger);
+        listeners.put(id, listener);
+        listener.next();
+        return id;
+    }
+
+    @Override
+    public void removeObserver(@Nonnull UUID registrationId) {
+        checkIfLive();
+        RingbufferListener<T> listener = listeners.remove(registrationId);
+        if (listener == null) {
+            throw new IllegalArgumentException(
+                    String.format("No registered observer with registration ID %s", registrationId));
+        } else {
+            listener.cancel();
+        }
+    }
+
+    @Override
+    public void setCapacity(int capacity) {
+        String ringbufferName = ringbufferName(name);
+        if (ringbufferExists(ringbufferName)) {
+            throw new IllegalStateException("Underlying buffer for observable '" + name + "' is already created.");
+        }
+        hzInstance.getConfig().addRingBufferConfig(new RingbufferConfig(ringbufferName).setCapacity(capacity));
+    }
+
+    @Override
+    public int getCapacity() {
+        String ringbufferName = ringbufferName(name);
+        if (ringbufferExists(ringbufferName)) {
+            return (int) hzInstance.getRingbuffer(ringbufferName).capacity();
+        } else {
+            //theoretically we could read the default config's capacity, but that's not possible
+            //when the HazelcastInstance is just a client
+            throw new IllegalStateException("Underlying buffer for observable '" + name + "' is not yet created.");
+        }
+    }
+
+    private boolean ringbufferExists(String ringbufferName) {
+        return hzInstance.getDistributedObjects().stream().anyMatch(
+                o -> o.getServiceName().equals(RingbufferService.SERVICE_NAME) && o.getName().equals(ringbufferName));
+    }
+
+    @Override
+    public void destroy() {
+        checkIfLive();
+
+        listeners.keySet().forEach(this::removeObserver);
+        listeners.clear();
+        listeners = null;
```

Comment:
We shouldn't incapacitate the instance. It's racy. We should make the instance quietly recreate resources if used again after `destroy`

Reply:
It was my idea, but I see that it's racy now (listeners is not `volatile`) and even then there will be race conditions. I guess it's fine, since it's the default behaviour for other HZ objects.

Reply:
Maybe calling `onDestroy` first and then continuing with others will be better. Also making the `listeners` volatile.